`List`: register the class with the `to_aiida_type` serializer

sphuber opened the pull request:
Fixes #5141 

This will allow users to pass a plain list as input for a port that
accepts `List` nodes and specified the `to_aiida_type` dispatch as
serializer. The value will be automatically converted to a `List` node.
This brings the functionality for `List` on par with the other base type
data node classes.

To make the newly added test work, the `Dict` class needed a change to allow comparing to normal dictionaries. This is added in a separate commit.

mbercx commented:
Nice! Love the fact that you don't have to worry about passing in a `List` anymore. However, won't this change the way that `Dict`s compare? I.e. won't:

```Python
Dict(dict={'a': 1}) == Dict(dict={'a': 1})
```

Now return `True`? I'm all for it, but I do think this behaviour is different from before (also see the discussion in https://github.com/aiidateam/aiida-core/issues/4495).

chrisjsewell reviewed (changes requested):
Yeh cheers, but just stick to the list addition for this PR

sphuber commented:
Forgot about that long open discussion about node equality. Agree that I shouldn't change that here. Still I think that the `assert dict_node == dict_node.value` should be non-controversial, right? I have updated the implementation to allow for this case, but to leave the original behavior when comparing node instances.

mbercx commented:
> Forgot about that long open discussion about node equality. Agree that I shouldn't change that here. Still I think that the `assert dict_node == dict_node.value` should be non-controversial, right? I have updated the implementation to allow for this case, but to leave the original behavior when comparing node instances.

I'm still in favour of actually making `Dict` nodes the same as `List` nodes (and other `BaseType`s) with regard to comparisons. ^^ Maybe not change that here, but we should try and hash this out before v2.0, perhaps? If we can get everyone to pitch in at the next AiiDA team meeting, I'm happy to open a PR to fix #4495 after, based on whatever we decide.

sphuber commented:
> I'm still in favour of actually making `Dict` nodes the same as `List` nodes (and other `BaseType`s) with regard to comparisons. ^^ 

I would have to revise the discussion once again and make a summary of the pro's and con's. I am tempted to agree with you but I remember there being strong con's that I cannot remember right now.

> Maybe not change that here, but we should try and hash this out before v2.0, perhaps?

Whether we change it or not, I think it would be good either way to come to a decision. So indeed good to try and make an executive summary and get a poll going with a final decision.

mbercx commented:
> Whether we change it or not, I think it would be good either way to come to a decision. So indeed good to try and make an executive summary and get a poll going with a final decision.

Alright, I'll try to go through the whole discussion again and write a summary by Wednesday. Then we can schedule a small meeting after the group meeting with everyone interested in discussing this (probably won't have enough time during the AiiDA meeting).

chrisjsewell commented:
Yeh I’m for whatever makes the user experience best, which is most likely comparing base types as there value. Just a pain, because there will probably be no way to deprecate in a nice way, and it will likely break someone’s code/plugin who has been relying on previous behaviour. I think perhaps a good way would be like sqlalchemy has done, you can add an environmental variable to turn on certain warnings. We could have this turn on a warning for any code that calls these `__eq__` methods, so people can quickly identify what code they need to change, then turn it off again

chrisjsewell commented:
(and this warning could be backported to v1

sphuber commented:
> Alright, I'll try to go through the whole discussion again and write a summary by Wednesday. Then we can schedule a small meeting after the group meeting with everyone interested in discussing this (probably won't have enough time during the AiiDA meeting).

Would be good, if you manage in time, to send the summary by email, such that those who are interested can have a look before. It is not really possible to do this during the meeting. Not that we should have the discussion during the meeting, but it can be helpful for people to have the context beforehand.

sphuber commented:
@chrisjsewell are there any objections left after I changed the `__eq__` implementation to not change comparison of nodes but only with plain dicts?

@mbercx would also be good to make sure #4495 gets in soon, since we likely also want to add this in 2.0 now that we are on a campaign to make all base types equal. Did you want to do this yourself, or can I also go ahead and implement it?

chrisjsewell commented:
Will have a double check soon, bare with me, I’m just in the middle of Croatia waiting for a bus lol

mbercx commented:
> @mbercx would also be good to make sure #4495 gets in soon, since we likely also want to add this in 2.0 now that we are on a campaign to make all base types equal. Did you want to do this yourself, or can I also go ahead and implement it?

I think I should find time to do this on Friday, or next week at the latest. Note that we should probably first agree on what the equality comparison between base types should be though, because making `List` and `Dict` `BaseType`s changes the equality comparison behavior for `Dict`.

sphuber commented:
@mbercx would addressing some of the feature requests really change the equality behavior? For example, adding the `value` property and changing the signature of the constructor wouldn't affect it, would it? Worst come to worst, we can always add these changes and leave the equality question for what it is if we cannot come to an agreement. In any case, we should probably change these things in separate commits if not also separate PRs.

sphuber commented:
> Will have a double check soon, bare with me, I’m just in the middle of Croatia waiting for a bus lol

pinging @chrisjsewell

chrisjsewell commented:
yep will do after lunch

chrisjsewell reviewed (approved):
Errr, I still don't feel the equality change needs to be present in this PR 😬
But anyhow, it's very likely it will be changed again before v2 releases, so I'll leave it up to you 🤷

sphuber commented:
> Errr, I still don't feel the equality change needs to be present in this PR grimacing

Well I guess it is not strictly necessary, but I don't think it hurts. And it allows the tests in the follow-up commit to use the same assert statement `assert converted == value` in the parametrized test. Without this addition, I would have to resort to making an exceptional case for the dictionary test.

chrisjsewell commented:
Oh I see, in `BaseType` there is already:

```python
    def __eq__(self, other):
        if isinstance(other, BaseType):
            return self.value == other.value
        return self.value == other
```

thats why `Dict` was previously different @sphuber 

This then raises the question, should `Dict` also inherit from `BaseType`?
Or, at a minimum, we should also implement the `__ne__` method on `Dict` as well

sphuber commented:
The reason I added the `__eq__` is indeed exactly because `Dict` doesn't inherit from `BaseType`. I didn't want to change the base class for `Dict` as that could have had even further consequences to the behavior and I explicitly wanted the change to be minimal and just have the node content compare equal to the base type of its value. I added the `self is other` as the fall-through comparison as that is the base implementation of `object`. However, I realize now, that I probably should have called `super().__eq__(self, other)` since `Node` defines a fall-through based on UUID. We should correct this therefore I think.

Then on whether `__ne__` should be specified: I don't think this is necessary. The [docs specify that](https://docs.python.org/3/reference/datamodel.html#object.__ne__) `__ne__` is automatically defined as the inverse of `__eq__`.

mbercx commented:
> @mbercx would addressing some of the feature requests really change the equality behavior? For example, adding the `value` property and changing the signature of the constructor wouldn't affect it, would it? Worst come to worst, we can always add these changes and leave the equality question for what it is if we cannot come to an agreement. In any case, we should probably change these things in separate commits if not also separate PRs.

Sorry, missed this reply! Will work on adding the `value` property and changing the constructor today.
